### PR TITLE
Fix and tweak how resource reload listeners are added

### PR DESCRIPTION
**Pull request by Cyborgmas**
From #6838, the new resource manager is DataPackRegistries. However instead of creating it once, it is created on each reload.

Not sure if my fix is proper/in line with how forge does things.

### CODE REVIEW

File: src/main/java/net/minecraftforge/resource/IForgeDataPackRegistries.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import net.minecraft.resources.IFutureReloadListener;
+import net.minecraftforge.common.loot.LootModifierManager;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+public interface IForgeDataPackRegistries {
+   List<IFutureReloadListener> resources = new ArrayList<>();
```
Comment (covers1624) [inline src/main/java/net/minecraftforge/resource/IForgeDataPackRegistries.java]: Because this is in an interface, this is public, it should not be.

**Review by covers1624 — COMMENTED** [review]
Why not just add an event to add things instead of this static 'registry'?

**Comment by Cyborgmas** [comment]
Yeah the list being public felt weird. What are the thoughts about this way with the Event?

**Comment by covers1624** [comment]
Yeah looks better this way, needs docs for when the event is fired, on what bus, and what its purpose is.

File: patches/minecraft/net/minecraft/resources/DataPackRegistries.java.patch [hunk 2]
```diff
@@ -1,28 +1,10 @@
...
  
     public FunctionReloader func_240960_a_() {
-@@ -85,4 +88,8 @@
-    public void close() {
-       this.field_240952_b_.close();
-    }
-+
-+   public net.minecraftforge.common.loot.LootModifierManager getLootModifierManager() {
-+      return lootManager;
-+   }
```
Comment (bignelli) [inline patches/minecraft/net/minecraft/resources/DataPackRegistries.java.patch:27]: Wasn't this getter used by Global Loot Modifiers ? I think they are broken right now though.

Comment (Cyborgmas) [reply patches/minecraft/net/minecraft/resources/DataPackRegistries.java.patch:27]: woops ill re add a getter

File: src/main/java/net/minecraftforge/common/ForgeMod.java [hunk 3]
```diff
@@ -23,8 +23,8 @@
 import net.minecraft.entity.ai.attributes.RangedAttribute;
 import net.minecraft.util.SoundEvent;
 import net.minecraft.world.storage.IServerConfiguration;
-import net.minecraft.world.storage.IWorldInfo;
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/common/ForgeMod.java]: Why are the imports in this file changing?

**Comment by Cyborgmas** [comment]
Fixed the things added docs. however not sure about how I handle getting the Instance in LootModifier. should be fine

File: src/main/java/net/minecraftforge/event/AddReloadListenerEvent.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The main ResourceManager is recreated on each reload, through {@link DataPackRegistries}'s creation.
+ *
+ * The event is fired on each reload and lets modders add their own ReloadListeners.
+ * The event is fired on the {@link MinecraftForge#EVENT_BUS}
+ */
+public class AddReloadListenerEvent extends Event
```
Comment (KnightMiner) [inline src/main/java/net/minecraftforge/event/AddReloadListenerEvent.java:18]: Might want to emphasize this is for server reload listeners. Only clue in the javadoc is the link to data pack registries, but I could see someone getting confused and trying to use this for the client side resource pack reload listeners.

**Comment by pupnewfster** [comment]
Brief summary of the core part of the issue (as requested by lex to remember), in 1.16 every time `/reload` is ran the reload listeners list is nuked and rebuilt. This PR solves the issue of mods not being able to reliably add reload listeners by adding an event to add listeners to that is fired when the reload listener list is being rebuilt.

File: src/main/java/net/minecraftforge/common/loot/LootModifierManager.java [hunk 5]
```diff
@@ -154,4 +161,17 @@ private IGlobalLootModifier deserializeModifier(ResourceLocation location, JsonE
         return registeredLootModifiers.values();
     }
 
+    @SubscribeEvent
+    public static void onResourceReload(AddReloadListenerEvent event)
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/loot/LootModifierManager.java]: Forge has its own internal handler, so this shouldn't have this weird meta instance thing...
What is the point of people having access to the instance?

Comment (Cyborgmas) [reply src/main/java/net/minecraftforge/common/loot/LootModifierManager.java]: used here https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/main/java/net/minecraftforge/common/ForgeHooks.java#L1169 but commented for now. I can put this in the InternalHandler if its better suited.

Comment (pupnewfster) [reply src/main/java/net/minecraftforge/common/loot/LootModifierManager.java]: Yes this should be moved to the internal handler as there is no sense in exposing a modder facing API if we can avoid it when it comes to internal instances of things.

**Comment by Cyborgmas** [comment]
Okay re-enabled global loot modifiers as well, this will close #6863

**Pull request merged**